File: backend/app/schemas/dna.py
from datetime import datetime
from typing import List

from pydantic import BaseModel


class DnaAnalyzeRequest(BaseModel):
    sequence: str


class DnaAnalyzeResponse(BaseModel):
    id: int
    sequence: str
    length: int
    gc_content: float
    melting_temp: float
    melting_temp_method: str
    molecular_weight: float
    rna: str
    protein: str
    markers: List[str]
    created_at: datetime


class DnaHistoryItem(BaseModel):
    id: int
    sequence: str
    length: int
    gc_content: float
    melting_temp: float
    created_at: datetime

File: frontend/src/types/dna.ts
/*
 * Mirrors backend/app/schemas/dna.py field-for-field. Manual sync point.
 *
 * Note `melting_temp_method` is returned alongside `melting_temp` so the API is
 * self-documenting about which Tm model produced the number (Biopython Tm_NN).
 */

export type DnaAnalysis = {
  id: number;
  sequence: string;
  length: number;
  gc_content: number;
  melting_temp: number;
  melting_temp_method: string;
  molecular_weight: number;
  /** Transcribed RNA. */
  rna: string;
  /** Translated protein; stop codons are normalized to `_` (not Biopython's `*`). */
  protein: string;
  markers: string[];
  created_at: string;
};

export type DnaHistoryItem = {
  id: number;
  sequence: string;
  length: number;
  gc_content: number;
  melting_temp: number;
  created_at: string;
};
